suprb/suprb.py

```diff
@@ -186,8 +193,27 @@ def _discover_rules(self, X: np.ndarray, y: np.ndarray, n_rules: int):
         if not self.pool_:
             warnings.warn(
                 "The population is empty, even after generating rules. "
-                "Solution optimization will be skipped.",
+                "Rule generation will be attempted another four times.",
                 PopulationEmptyWarning)
+            # Needed to avoid errors about solution being empty/having no fitness
+            for i in range(4):
+                new_rules = self.rule_generation_.optimize(X, y, n_rules=n_rules)
+                if new_rules:
+                    warnings.warn(
+                        f"It took another {i + 1} discovery-phases to find suitable rules. ",
+                        PopulationEmptyWarning)
+                    self.pool_.extend(new_rules)
+                    break
```

Comment:
You can move this whole block into a separate function something like `retry_rule_discovery`, then return if new rules are found (instead of break)